### PR TITLE
Add processing of target inputs in decorator task

### PR DESCRIPTION
**Pull request by EzzhevNikita**
This PR is a part of the feature for injecting task input into the decorator's tasks. For security reasons, we should filter out inputs that contain secrets. If the task injected by the decorator is trying to access input with a secret, it will be skipped.

Tested with injected tasks that are trying to pick up inputs with runtime secrets (that were set thought vso command)

### CODE REVIEW

**Review by tatyana-kostromskaya — APPROVED** [review]

File: src/Agent.Worker/TaskRunner.cs [hunk 1]
```diff
@@ -525,5 +538,24 @@ private void PrintTaskMetaData(Definition taskDefinition)
             ExecutionContext.Output($"Help         : {taskDefinition.Data.HelpUrl ?? taskDefinition.Data.HelpMarkDown}");
             ExecutionContext.Output("==============================================================================");
         }
+
+        private bool IsInputContainsSecret(string inputValue)
+        {
+            string maskedString = HostContext.SecretMasker.MaskSecrets(inputValue);
+            return maskedString.Contains("***");
```
Comment (anatolybolshakov) [inline src/Agent.Worker/TaskRunner.cs:545]: Nit: this creates indirect duplication (you rely on fact that secrets masked by ***) - could there be any more generic way to check it? Could inputValue != maskedString condition be more generic here?

File: src/Agent.Worker/TaskRunner.cs [hunk 2]
```diff
@@ -525,5 +538,24 @@ private void PrintTaskMetaData(Definition taskDefinition)
...
+        }
+
+        private bool ValidateInjectedTaskInputs(Dictionary<string, string> inputs)
```
Comment (anatolybolshakov) [inline src/Agent.Worker/TaskRunner.cs:548]: Could you please make the name more specific? Basically it simply checks if inputs contain any secrets - although validation is something more generic

File: src/Agent.Worker/TaskRunner.cs [hunk 3]
```diff
@@ -209,6 +209,19 @@ public async Task RunAsync()
...
                 runtimeVariables.ExpandValues(target: inputs);
+
+                if (Task.Name.StartsWith("__system_posttargettask_")
+                    || Task.Name.StartsWith("__system_pretargettask_"))
+                {
+                    bool isValid = this.ValidateInjectedTaskInputs(inputs);
+                    if (!isValid)
+                    {
+                        ExecutionContext.Result = TaskResult.Skipped;
+                        ExecutionContext.ResultCode = $"It is not allowed to pass inputs that contain secrets to the tasks injected by decorators";
```
Comment (anatolybolshakov) [inline src/Agent.Worker/TaskRunner.cs:220]: Can we share target task here for which it failed?

Comment (EzzhevNikita) [reply src/Agent.Worker/TaskRunner.cs]: Added logic for printing input names that contains secrets

File: src/Agent.Worker/TaskRunner.cs [hunk 4]
```diff
@@ -209,6 +209,23 @@ public async Task RunAsync()
                 // Expand the inputs.
                 Trace.Verbose("Expanding inputs.");
                 runtimeVariables.ExpandValues(target: inputs);
+
+                if (Task.Name.StartsWith("__system_posttargettask_")
+                    || Task.Name.StartsWith("__system_pretargettask_"))
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/TaskRunner.cs:214]: Could it make sense to export `__system_pretargettask_` and `__system_pretargettask_` as constants? Or add some field to `Task` that will allow us to determine if the task was injected from the decorator? 

Also what about other decorator's targets? I think we also should not provide access to secret variables for them as well. Or am I missing something?

File: src/Agent.Worker/TaskRunner.cs [hunk 5]
```diff
@@ -525,5 +542,25 @@ private void PrintTaskMetaData(Definition taskDefinition)
             ExecutionContext.Output($"Help         : {taskDefinition.Data.HelpUrl ?? taskDefinition.Data.HelpMarkDown}");
             ExecutionContext.Output("==============================================================================");
         }
+
+        private bool ContainsSecret(string inputValue)
+        {
+            string maskedString = HostContext.SecretMasker.MaskSecrets(inputValue);
+            return maskedString != inputValue;
+        }
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/TaskRunner.cs:550]: Could you please add some comments to this method? Also, maybe it makes sense to implement the `IsContain` method for the `SecretMasker` class?

File: src/Agent.Worker/TaskRunner.cs [hunk 6]
```diff
@@ -525,5 +542,25 @@ private void PrintTaskMetaData(Definition taskDefinition)
...
+            foreach (var input in inputs)
+            { 
+                if (input.Key.StartsWith("target_") && this.ContainsSecret(input.Value))
+                {
+                    inputsWithSecrets.Add(input.Key);
+                }
+            }
+
+            return inputsWithSecrets;
+        }
```
Comment (alexander-smolyakov) [inline src/Agent.Worker/TaskRunner.cs:564]: Could you please add some comments to this method?

File: src/Agent.Worker/TaskRunner.cs [hunk 7]
```diff
@@ -209,6 +209,23 @@ public async Task RunAsync()
...
+                    if (inputsWithSecrets.Count > 0)
+                    {
+                         string inputsForReport = string.Join(Environment.NewLine,
+                             inputsWithSecrets.Select(input => string.Join("\n", input)));
+
+                        ExecutionContext.Result = TaskResult.Skipped;
+                        ExecutionContext.ResultCode = $"Task is trying to access these inputs, that contains secrerts:\n{inputsForReport}\nIt is not allowed to pass inputs that contain secrets to the tasks injected by decorators.";
+                        return;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (EzzhevNikita) [reply src/Agent.Worker/TaskRunner.cs]: Yes moved it to _CheckInjectedTasksInputs_ method

File: src/Agent.Worker/TaskRunner.cs (hunk 4 above)
Comment (EzzhevNikita) [reply src/Agent.Worker/TaskRunner.cs]: I have created an internal static class - _InjectedTasksNamesPrefixes_ to store these constants

**Review by alexander-smolyakov — APPROVED** [review]
LGMT, thanks! 
Also, tested these changes locally all works fine.

File: src/Misc/layoutbin/en-US/strings.json [hunk 8]
```diff
@@ -560,6 +560,7 @@
   "ScanToolCapabilities": "Scanning for tool capabilities.",
   "ScreenSaverPoliciesInspection": "Checking for policies that may prevent screensaver from being disabled.",
   "ScreenSaverPolicyWarning": "Screensaver policy is defined on the machine. This may lead to screensaver being enabled again. Active screensaver may impact UI operations, for e.g., automated UI tests may fail.",
+  "SecretsAreNotAllowedInInjectedTaskInputs": "Task is trying to access these inputs, that contains secrerts:\n{0}\nIt is not allowed to pass inputs that contain secrets to the tasks injected by decorators.",
```
Comment (anatolybolshakov) [inline src/Misc/layoutbin/en-US/strings.json:563]:
```suggestion
  "SecretsAreNotAllowedInInjectedTaskInputs": "Task is trying to access the following inputs of a target task which contain secrets:\n{0}\nIt is not allowed to pass inputs that contain secrets to the tasks injected by decorators.",
```

File: src/Agent.Worker/TaskRunner.cs [hunk 9]
```diff
@@ -525,5 +535,74 @@ private void PrintTaskMetaData(Definition taskDefinition)
             ExecutionContext.Output($"Help         : {taskDefinition.Data.HelpUrl ?? taskDefinition.Data.HelpMarkDown}");
             ExecutionContext.Output("==============================================================================");
         }
+
+        /// <summary>
+        /// Used to check if provided input value contain any secret
+        /// </summary>
+        /// <param name="inputValue">Value of input to check</param>
+        /// <returns>Returns `true` if provided string contain secret, otherwise `false`</returns>
+        private bool ContainsSecret(string inputValue)
```
[the remark this answers is not in the document]
Comment (EzzhevNikita) [reply src/Agent.Worker/TaskRunner.cs]: Moved all decorator-related methods to the TaskDecoratorManager manager.

File: src/Agent.Worker/TaskRunner.cs [hunk 10]
```diff
@@ -525,5 +535,74 @@ private void PrintTaskMetaData(Definition taskDefinition)
...
+            return inputsWithSecrets;
+        }
+
+        /// <summary>
+        /// Used to check list of inputs in injected task that started with target_ prefix and contain secrets,
+        /// such inputs are autoinjected from target tasks
+        /// </summary>
+        /// <param name="inputs">Inputs presented as a dictionary with input name as key and input's value as the value of the corresponding key</param>
+        /// <returns>Returns list of inputs' names that contain secret values</returns>
+        private void CheckInjectedTasksInputs(Dictionary<string, string> inputs)
```
[the remark this answers is not in the document]
Comment (EzzhevNikita) [reply src/Agent.Worker/TaskRunner.cs]: Changed the logic of this method and renamed it.

File: src/Agent.Worker/TaskRunner.cs [hunk 11]
```diff
@@ -209,6 +209,22 @@ public async Task RunAsync()
                 // Expand the inputs.
                 Trace.Verbose("Expanding inputs.");
                 runtimeVariables.ExpandValues(target: inputs);
+
+                // We need to verify inputs of the tasks' that were injected by decorators, to check if they contain secrets,
+                // for security reasons execution of tasks' in this case should be skipped.
+                // Target task inputs could be injected into the decorator's tasks if the decorator has post-task-tasks or pre-task-tasks targets,
+                // such tasks will have names that start with __system_pretargettask_ or __system_posttargettask_.
+                var taskDecoratorManager = HostContext.GetService<ITaskDecoratorManager>();
+                if (taskDecoratorManager.IsInjectedTaskForTarget(Task.Name) &&
```
Comment (anatolybolshakov) [inline src/Agent.Worker/TaskRunner.cs:218]: Could the Task.Name be null?

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!

**Pull request merged**